### DISCUSSION TITLE
FluxCD v2 not deploying new image automatically.

### DISCUSSION THREAD

**Discussion by evanrich** [fluxcd/flux2#1730] [Q&A] [answered]
I tried to migrate from v1 to v2, but had too many issues, so I deleted the fluxv1 namespace, and bootstrapped v2.  After a few hours, I was able to get pretty much everything set up, including having flux push changes to my git repo, but it's not ultimately re-deploying containers with new images.

My repo/local path looks like this:
```
/srv/homek8s/
├── clusters
│   └── my-cluster
│       ├── automation
│       │   ├── homek8s-auto.yaml
│       │   ├── jackett-image.yaml
│       │   ├── jackett-policy.yaml
│       │   ├── sonarr-image.yaml
│       │   └── sonarr-policy.yaml
│       └── flux-system
│           ├── gotk-components.yaml
│           ├── gotk-sync.yaml
│           └── kustomization.yaml
├── database
│   ├── influxdb
│   │   ├── 01networkpolicy.yaml
│   │   └── deployment.yaml
│   └── postgres
│       ├── 01networkpolicy.yaml
│       ├── postgresql-nextcloud-secret.yaml
│       ├── servicemonitor.yaml
│       └── sts.yaml
```

Basically, flux stuff is in /srv/homek8s/clusters/my-cluster/*, and then application manifests are in /srv/homek8s/database, /srv/homek8s/media, etc.

here's all my info:
```
 flux get sources all
NAME                            READY   MESSAGE                                                                 REVISION                                        SUSPENDED
gitrepository/flux-system       True    Fetched revision: master/796c1094b9d7a3930b8c4b6fe776fe3c016b6d59       master/796c1094b9d7a3930b8c4b6fe776fe3c016b6d59 False
````
```
 flux get image repository -A
NAMESPACE       NAME            READY   MESSAGE                                 LAST SCAN                       SUSPENDED
flux-system     jackett-image   True    successful scan, found 8796 tags        2021-08-18T06:19:25-07:00       False
flux-system     sonarr-image    True    successful scan, found 3420 tags        2021-08-18T10:17:08-07:00       False
```
```
flux get image policy -A
NAMESPACE       NAME            READY   MESSAGE                                                                         LATEST IMAGE
flux-system     jackett-policy  True    Latest image tag for 'linuxserver/jackett' resolved to: version-v0.18.551       linuxserver/jackett:version-v0.18.551
flux-system     sonarr-policy   True    Latest image tag for 'linuxserver/sonarr' resolved to: develop-3.0.6.1309-ls193 linuxserver/sonarr:develop-3.0.6.1309-ls193
```

and finally, ImageUpdateAutomation
```
flux get image update -A
NAMESPACE       NAME            READY   MESSAGE                                                         LAST RUN                        SUSPENDED
flux-system     homek8s-auto    True    no updates made; last commit c7b6640 at 2021-08-18T16:39:04Z    2021-08-18T11:44:36-07:00       False
```

After annotating a sts.yaml file, i can see that flux correctly changes the build, and commits it to git:
![image](https://user-images.githubusercontent.com/518320/129955114-79b6dfe3-e2eb-411d-8a1f-0b58d09c30e7.png)

but the container in `/srv/homek8s/media/jackett/sts.yaml` is never updated.  Do the application yaml's need to be in the cluster path?

**Comment by kingdonb** [comment, marked as answer]
Thanks for posting! Sorry that you are having trouble with this.

The annotated Flux YAMLs that `ImageUpdateAutomation` is going to update need to be present in the `spec.path` listed there, in Image Automation, relative to the root of the gitrepository that ImageUpdateAutomation lists in its `spec.sourceRef`.

It looks like your image policies are functioning and your image update is connected successfully and has cloned the git repo, but has no update. You might be missing updates because your ImageRepo is not reconciling; it is possible to make this faster using the Webhook Receiver for ImageRepo, if you control those image repos and can set webhooks on them.

What exactly is the change in that commit you got out of Flux Image Automation though, if it did not update the container? What comment did you use to annotate the image string in the `sts.yaml` file, and where is it located in the YAML manifest that you have automated? (Is the manifest properly inside of the path from `spec.path` I mentioned? Can we see the ImageUpdateAutomation definition that you have used?)

The examples for everyone who might find this question are at https://fluxcd.io/docs/guides/image-update/ – I assume you found them already.

**Reply by evanrich** [reply]
Thanks for the response @kingdonb   When I run a flux reconcile image update homek8s-auto, i get the following
```
flux reconcile image update homek8s-auto
► annotating ImageUpdateAutomation homek8s-auto in flux-system namespace
✔ ImageUpdateAutomation annotated
◎ waiting for ImageUpdateAutomation reconciliation
✔ ImageUpdateAutomation reconciliation completed
✔ no updates made; last commit f264d42 at 2021-08-18T23:03:43Z

```

When I do a git pull (answering your question), i see it changed stuff:
```
Updating c4dafeb..f264d42
Fast-forward
 media/tautulli/sts.yaml | 107 +++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------
 1 file changed, 53 insertions(+), 54 deletions(-)
```
It always comes back with "no changes made".  I can tell it does stuff, but it seems like it's rewriting the yaml, along with actually updating the image...take this for example:
![image](https://user-images.githubusercontent.com/518320/129983705-99854689-50d2-4d88-9b8b-94fa950f0701.png)

so it seems like flux is doing literally everything it should, except `kubectl apply -f <yaml.yaml>`

I hate to say it, but V1 just worked lol,  I'm probably missing something really dumb here, but the docs arent' well written.  Also, I dont control the repos, they're public ones (gitlab, linuxserver, etc).

Oh, and the annotation I'm using to get it to update the yaml is:
```
 image: linuxserver/tautulli:version-v2.7.5 # {"$imagepolicy": "flux-system:tautulli"}
```
so this works, flux re-writes the deployment.yaml with the correct new versions, it  commits the new .yaml to git, it just doesn't re-deploy.   If I understand you correctly, it expects the path to be relevant to the cluster path, so i think, for example, it should be:

```
/srv/homek8s/
├── clusters
│   └── my-cluster
         └──  media
│            └── jackett
│                  └── sts.yaml
│            └── sonarr
│                 └──sts.yaml
│       ├── automation
│       │   ├── homek8s-auto.yaml
│       │   ├── jackett-image.yaml
│       │   ├── jackett-policy.yaml
│       │   ├── sonarr-image.yaml
│       │   └── sonarr-policy.yaml
│       └── flux-system
│           ├── gotk-components.yaml
│           ├── gotk-sync.yaml
│           └── kustomization.yaml
```

basically the "namespaces" need to be nested under the cluster folder, yes?
Thanks!

**Reply by evanrich** [reply]
@kingdonb  I kinda just answered my own question.  I moved one of the services into the folder, so that it was 
```
clusters->mycluster->media->tautulli
```
and as soon as I ran a reconcile, the pod terminated and redeployed, as expected.  I suppose I can easily just do a `git mv` on everything to move it into the correct path.

Thanks!

**Reply by kingdonb** [reply]
I think you got it. Flux will only apply resources that `kustomize build --autodetect --recursive` would pick up, within the bounds of these parameters (which are broader than what you get by default with kustomize out of the box, load-restrictor=None for example):

* https://fluxcd.io/docs/components/kustomize/kustomization/#generate-kustomizationyaml
and
* https://fluxcd.io/docs/faq/#what-is-the-behavior-of-kustomize-used-by-flux

The docs are still somewhat work-in-progress and we have a sort of task-force working on redesigning them to make tasks more discoverable and to make it more straightforward to understand specific things without so much monolithic tutorial knowledge that you have to kind-of absorb all at once in somewhat superhuman fashion, unfortunately there is a lot to know about Kustomize and getting Flux users exposed to all the details in an orderly fashion is still one of the harder parts about working with Flux.

Glad to hear you got this sorted out!
